File: codeforces/BinaryStringMinimizer.hs
-- http://codeforces.com/problemset/problem/1256/D

import Data.List


zeroChar = '0'
oneChar = '1'


data Case = Case { moves  :: Int
                 , string :: String }


getMoves :: String -> Int
getMoves string = read $ last $ words string :: Int


toCase :: (String, String) -> Case
toCase (one, two) = Case { moves = getMoves one, string = two }


toPairs :: [a] -> [(a,a)]
toPairs []                  = []
toPairs [_]                 = error "list size is odd"
toPairs (first:second:tail) = (first, second) : toPairs tail


calculateIndicesAfterMoves :: [Int] -> [Int] -> Int -> [Int]
calculateIndicesAfterMoves [] _ _ = []
calculateIndicesAfterMoves initialZeros _ 0 = initialZeros
calculateIndicesAfterMoves (initialZero:initialZeros) (idealIndex:idealIndices) moves =
  newZeroIndex : calculateIndicesAfterMoves initialZeros idealIndices updatedMoves
  where desiredMoves = initialZero - idealIndex
        allowedMoves = min desiredMoves moves
        updatedMoves = moves - allowedMoves
        newZeroIndex = initialZero - allowedMoves


buildFinalList :: [Int] -> Int -> String -> String
buildFinalList [] index ones = ones
buildFinalList (zero:zeros) index (one:ones)
  | zero == index = zeroChar : buildFinalList zeros (index + 1) ones
  | otherwise     = one : buildFinalList (zero:zeros) (index + 1) ones


solve :: Case -> String
solve (Case moves string) = do
  let initialZeroIndices = elemIndices zeroChar string
  let idealIndices      = [0..(length initialZeroIndices)]
  let finalZeroIndices   = calculateIndicesAfterMoves initialZeroIndices idealIndices moves
  let ones               = replicate (length string) oneChar
  buildFinalList finalZeroIndices 0 ones


main :: IO ()
main = interact $
  unlines . map (id . solve . toCase) . -- Why id? See: https://stackoverflow.com/a/12104586/2420718
  toPairs . tail . lines

File: codeforces/binary_string_minimizer.py
class Case:
    def __init__(self, size, moves, string):
        self.size = size
        self.moves = moves
        self.string = string


SPACE = ' '
ZERO = '0'
ONE = '1'


def get_case():
    line_elements = input().split(SPACE)
    size = int(line_elements[0])
    moves = int(line_elements[1])
    string = list(input())
    return Case(size, moves, string)


def get_zero_indices(string):
    zero_indices = []
    for index, char in enumerate(string):
        if char == ZERO:
            zero_indices.append(index)
    return zero_indices


def get_zero_indices_after_moves(initial_zero_indices, ideal_indices, moves):
    final_zero_indices = []
    if len(initial_zero_indices) != len(ideal_indices):
        raise ValueError('initial_zero_indices and ideal_indices should be of '
                         'the same length')
    for i, zero_index in enumerate(initial_zero_indices):
        if moves < 0:
            raise ValueError('moves should be >= 0')
        if moves == 0:
            final_zero_indices.extend(initial_zero_indices[i:])
            return final_zero_indices
        desired_moves = zero_index - ideal_indices[i]
        allowed_moves = min(moves, desired_moves)
        new_zero_index = zero_index - allowed_moves
        final_zero_indices.append(new_zero_index)
        moves = moves - allowed_moves
    return final_zero_indices


def add_zeros(ones, zero_indices):
    for index in zero_indices:
        ones[index] = ZERO
    return ones


def solve(case):
    initial_zero_indices = get_zero_indices(case.string)
    ideal_zero_indices = list(range((len(initial_zero_indices))))
    final_zero_indices = get_zero_indices_after_moves(
        initial_zero_indices, ideal_zero_indices, case.moves)
    ones = [ONE] * case.size
    response = add_zeros(ones, final_zero_indices)
    return response


if __name__ == "__main__":
    number_of_cases = int(input())
    for i in range(number_of_cases):
        case = get_case()
        solvedList = solve(case)
        solution = "".join(solvedList)
        print(solution)
